src/metadata_encoding.rs

```diff
@@ -0,0 +1,203 @@
+use bytes::Bytes;
+use http::header::HeaderValue;
+use std::fmt;
+use std::error::Error;
+use std::hash::Hash;
+
+/// A possible error when converting a `MetadataValue` from a string or byte
+/// slice.
+#[derive(Debug, Default)]
+pub struct InvalidMetadataValue {
+    _priv: (),
+}
+
+pub trait ValueEncoding: Clone + Eq + PartialEq + Hash {
```

Comment:
Is this meant to be a "sealed" internal trait? As in, only types within this crate implement it, and users of the crate don't actually ever touch the trait? If so, there's some Rust type system trickery that we can do to make the trait less "public":

- Add an internal private module `sealed`, with a `pub trait Sealed` inside it.
- Add `self::sealed::Sealed` as a super trait of `ValueEncoding`.
- Implement `Sealed` for `Ascii` and `Binary`.
- Optionally move all the `#[doc(hidden)]` methods to the `Sealed` trait, so no one can even call them. This allows changing their signature without breaking external code.

Reply:
Thanks for the info on how to do this, I will update the PR when I get time.